rust/chains/hyperlane-cosmos/src/mailbox/contract.rs

```diff
@@ -0,0 +1,246 @@
+use std::fmt::{Debug, Formatter};
+use std::num::NonZeroU64;
+use std::str::FromStr;
+
+use async_trait::async_trait;
+use cosmrs::proto::cosmos::base::abci::v1beta1::TxResponse;
+use tracing::{instrument, warn};
+
+use hyperlane_core::{
+    utils::bytes_to_hex, ChainResult, ContractLocator, HyperlaneChain, HyperlaneContract,
+    HyperlaneDomain, HyperlaneMessage, HyperlaneProvider, Mailbox, RawHyperlaneMessage,
+    TxCostEstimate, TxOutcome, H256, U256,
+};
+
+use crate::address::CosmosAddress;
+use crate::grpc::WasmProvider;
+use crate::payloads::general;
+use crate::payloads::mailbox::{
+    GeneralMailboxQuery, ProcessMessageRequest, ProcessMessageRequestInner,
+};
+use crate::types::tx_response_to_outcome;
+use crate::utils::get_block_height_for_lag;
+use crate::{payloads, ConnectionConf, CosmosProvider, Signer};
+
+#[derive(Clone)]
+/// A reference to a Mailbox contract on some Cosmos chain
+pub struct CosmosMailbox {
+    config: ConnectionConf,
+    domain: HyperlaneDomain,
+    address: H256,
+    provider: CosmosProvider,
+}
+
+impl CosmosMailbox {
+    /// Create a reference to a mailbox at a specific Ethereum address on some
+    /// chain
+    pub fn new(
+        conf: ConnectionConf,
+        locator: ContractLocator,
+        signer: Option<Signer>,
+    ) -> ChainResult<Self> {
+        let provider = CosmosProvider::new(
+            locator.domain.clone(),
+            conf.clone(),
+            Some(locator.clone()),
+            signer,
+        )?;
+
+        Ok(Self {
+            config: conf,
+            domain: locator.domain.clone(),
+            address: locator.address,
+            provider,
+        })
+    }
+
+    /// Prefix used in the bech32 address encoding
+    pub fn bech32_prefix(&self) -> String {
+        self.config.get_bech32_prefix()
+    }
+
+    fn contract_address_bytes(&self) -> usize {
+        self.config.get_contract_address_bytes()
+    }
+}
+
+impl HyperlaneContract for CosmosMailbox {
+    fn address(&self) -> H256 {
+        self.address
+    }
+}
+
+impl HyperlaneChain for CosmosMailbox {
+    fn domain(&self) -> &HyperlaneDomain {
+        &self.domain
+    }
+
+    fn provider(&self) -> Box<dyn HyperlaneProvider> {
+        Box::new(self.provider.clone())
+    }
+}
+
+impl Debug for CosmosMailbox {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        todo!()
```

Comment:
lol - also realize this already existed but this feels dangerous, could you fix this as part of your efforts? doesn't need to be in this PR

Could definitely see us accidentally logging the Mailbox and getting a panic 😅😭

Reply:
Fixed.